src/riak_kv_query_server.erl

```diff
@@ -0,0 +1,678 @@
+%% -------------------------------------------------------------------
+%%
+%% riak_query_worker: Manage complex secondary index query.
+%%
+%% Copyright (c) 2007-2013 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc The query worker manages a single secondary index query.
+%%
+%% This is a gen_server to manage an individual query.
+%% 
+%% The server will initialise a coverage query, and then await vnodes
+%% returning results.  The server should terminate when:
+%% - all vnodes have returned all results
+%% - sufficient results from all vnodes have been received such that any
+%% further results received would exceed the maximum number of results
+%% - the PID which requested the query dies
+%% - an absolute timeout occurs
+%% 
+%% TODO:
+%% This has not been implemented with the riak_core_coverage_fsm behaviour as
+%% it was considered there was minimal value in that behaviour, and there is a
+%% preference not to complicate future work to remove deprecated gen_fsm
+%% from Riak by expanding its use.
+%% In the future the riak_kv_clusteraae_fsm and riak_index_fsm should be
+%% refactored in line with this - and this may involve re-introducing a
+%% common behaviour.
+
+-module(riak_kv_query_server).
+
+-behaviour(gen_server).
+
+-define(SLOW_TIME, application:get_env(riak_kv, index_fsm_slow_timems, 200)).
+-define(FAST_TIME, application:get_env(riak_kv, index_fsm_fast_timems, 10)).
+
+-export(
+    [
+        query_start/1,
+        new_timeout/2,
+        check_vnode_monitor/1
+    ]
+).
+
+-export(
+    [
+        init/1,
+        handle_call/3,
+        handle_cast/2,
+        handle_info/2,
+        terminate/2,
+        code_change/3,
+        format_status/1
+    ]
+).
+
+-include_lib("kernel/include/logger.hrl").
+
+-define(VERSION, [{version, 2}]). % to be used in sets
+-define(ETS_THRESHOLD, 4096).
+% -define(START_OPTS, [{spawn_opt, [{min_heap_size, ?MIN_HEAP_SIZE}]}]).
```

Comment:
Is this a left over from tests, or for experiments. Add a comment why this is an alternative or leave it out completely.